File: Online Judges/LeetCode/May LeetCoding Challenge/Week-4/Possible Bipartition.py
class Solution:
    def possibleBipartition(self, N: int, dislikes: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for u, v in dislikes:
            graph[v].append(u)
            graph[u].append(v)

        color = {}

        def dfs(u, c=0):
            if u in color:
                return color[u] == c
            color[u] = c
            ret = 1
            for v in graph[u]:
                ret = ret & dfs(v, c ^ 1)
            return ret
        ans = 1
        for u in range(1, N+1):
            if u not in color:
                ans &= dfs(u, 0)
        return ans

File: Online Judges/LeetCode/May LeetCoding Challenge/Week-4/Possible Bipartition.cpp
class Solution
{
public:
    vector<int> graph[2001];
    int color[2001];
    bool dfs(int u, int c = 0)
    {
        if (color[u] != -1)
            return color[u] == c;
        color[u] = c;
        bool ret = 1;
        for (auto v : graph[u])
        {
            ret &= dfs(v, c ^ 1);
        }
        return ret;
    }
    bool possibleBipartition(int N, vector<vector<int>> &dislikes)
    {
        memset(color, -1, sizeof color);
        for (auto dis : dislikes)
        {
            int u = dis[0];
            int v = dis[1];
            graph[u].push_back(v);
            graph[v].push_back(u);
        }
        int Ans = 1;
        for (int i = 1; i <= N; i++)
        {
            if (color[i] == -1)
            {
                Ans &= dfs(i, 0);
            }
        }
        for (int i = 1; i <= N; i++)
            graph[i].clear();
        return Ans;
    }
};